db/core/htrie.c

```diff
@@ -792,22 +977,158 @@ tdb_htrie_insert(TdbHdr *dbh, unsigned long key, void *data, size_t *len)
 	goto retry;
 }
 
+static inline bool
+tdb_eval_eq_cb(tdb_eq_cb_t *eq_cb, TdbRec *rec, void *data)
+{
+	return ((eq_cb && eq_cb(rec, data)) || !eq_cb);
+}
+
+static inline void
+tdb_eval_bf_remove_cb(tdb_before_remove_cb_t *bf_remove_cb, TdbRec *rec)
+{
+	if (bf_remove_cb)
+		bf_remove_cb(rec);
+}
+
+#define TDB_REMOVE_FOREACH_REC(body)					\
+	do {								\
+		size_t rlen = sizeof(*rec) +				\
```

Comment:
In my understanding `TDB_HTRIE_RBODYLEN` is about the record _body_ length, not about the difference between `sizeof(TdbVRec)` and `sizeof(TdbFRec)`.

I think we have the bug. See that `tdb_htrie_create_rec()` assigns the data length to `rec->len`, so `TDB_HTRIE_RBODYLEN` for `TdbVRec` is the data length, carried by the record. Maybe the bug doesn't appear (frequently) since it affects only chained records, there is only 8 byte difference (which can be even smaller due to `TDB_HTRIE_RALIGN` - we simply may observe incorrect record instead of accessing invalid memory location). @EvgeniiMekhanik please also have a look - this bug may appear on the current pilot!

Reply:
I see, you are right we take the size of `TdbRec` even for `TdbVRec`. I fixed it throughout htrie code.